torch/ao/quantization/fx/_model_report/detector.py

```diff
@@ -577,17 +600,214 @@ def determine_observer_insert_points(self, prepared_fx_model: GraphModule) -> Di
+                    self.PER_CHANNEL_MAX_KEY: pre_obs.max_val,
+                    self.PER_CHANNEL_MIN_KEY: pre_obs.min_val,
+                    self.GLOBAL_MAX_KEY: max(pre_obs.max_val),
+                    self.GLOBAL_MIN_KEY: min(pre_obs.min_val),
+                }
+
+        return input_info
+
+    def _extract_weight_info(self, model: GraphModule) -> Dict[str, Dict]:
+        r"""
+        Takes in a callibrated GraphModule and then finds the relavent observers.
+        It then extracts the weight information for each layer an observer is attached to.
+
+        Args
+            model (GraphModule): The prepared and calibrated GraphModule with inserted ModelReportObservers
+
+        Returns a dict mapping module fqns (str) to a dict with keys:
+            "per_channel_max" : maps to the per_channel max values
+            "per_channel_min" : maps to the per_channel min values
+            "global_max" : maps to the global max recorded
+            "global_min" : maps to the global min recorded
+        """
+        # return dictionary mapping observer fqns to desired info
+        weight_info: Dict[str, Dict] = {}
+
+        for fqn, module in model.named_modules():
+            # check to see if module is of a supported type
+            is_supported_type: bool = (
+                sum(list(map(lambda x: type(module) is x, self.SUPPORTED_MODULES))) > 0
+            )
+
+            # if module is supported and it has a pre-observer
+            if is_supported_type and hasattr(module, self.DEFAULT_PRE_OBSERVER_NAME):
+                # we don't need actual observer, just the module weights
+                # calculate min and max vals
+                min_val, max_val = torch.aminmax(module.weight, dim=self.ch_axis)
+
+                # flatten entries since conv can have multiple dimensions
+                min_val = torch.flatten(min_val)
+                max_val = torch.flatten(max_val)
+
+                weight_info[fqn] = {
+                    self.PER_CHANNEL_MAX_KEY: max_val,
+                    self.PER_CHANNEL_MIN_KEY: min_val,
+                    self.GLOBAL_MAX_KEY: max(max_val),
+                    self.GLOBAL_MIN_KEY: min(min_val),
+                }
+
+        return weight_info
+
+    def _calculate_range_ratio(self, info_dict: Dict) -> torch.Tensor:
+        r"""
+        Takes in an info dict and calculates the s_c matrix.
+
+        Args:
+            info_dict (dict): A dictionary of either input or weight range info
+
+        Returns a tensor of values, where each value is the s_c stat for a different channel
+        """
+        # calculate the ratios of the info
+        per_channel_range = info_dict[self.PER_CHANNEL_MAX_KEY] - info_dict[self.PER_CHANNEL_MIN_KEY]
+        global_range = info_dict[self.GLOBAL_MAX_KEY] - info_dict[self.GLOBAL_MIN_KEY]
+
+        # if global range is 0, throw error
+        if global_range == 0:
+            raise ValueError("The range of the info dict is 0")
```

Comment:
i'd not reference the info dict in a user facing message, i.e. user has an input of all zeros, runs model report generator, we should probably give a better error message. Could just return None and handle it at upper level if that makes more sense.